### ISSUE TITLE
Downgrade warning is too agressive

### ISSUE THREAD

**Issue by ericstj** [NuGet/Home#948]
I'm getting warnings for downgrade even when nearest wins already chose a higher version.  Is that expected?

```
{
  "supports": {
    "net46.app": { },
    "uwp.10.0.app": { }
  },
  "dependencies": {
    "Microsoft.NETCore": "5.0.0-beta-23101",
    "Microsoft.NETCore.Portable.Compatibility": "1.0.0-beta-23101",
    "System.ServiceModel.Http": "4.0.10-beta-23101"
  },
  "frameworks": {
    "dotnet": { }
  }
}
```
Results in a downgrade warning 
Detected package downgrade: System.Runtime from 4.0.20-beta-23101 to 4.0.0-beta-23101 
 Project/ClassLibrary1 [1.0.0-beta-23101, ) -> Package/System.ServiceModel.Http [4.0.10-beta-23101, ) -> Package/System.Net.WebHeaderCollection [4.0.0-beta-23101, ) -> System.Runtime [4.0.20-beta-23101, ) 
 Project/ClassLibrary1 [1.0.0-beta-23101, ) -> Package/System.ServiceModel.Http [4.0.10-beta-23101, ) -> Package/System.Runtime [4.0.0-beta-23101, )			0

I understand that the package itself has an issue, but it's not an issue for the consumer of the package.  We can have this type of warning at pack time but it's just noise for the developer since it isn't causing any problem for their app.

**Comment by ericstj** [comment]
This same thing is causing "min version" PCLs to be extremely noisy.  I build a PCL against a an old portable profile and create a "dotnet" based package that can run on old and new.  For example: System.Reflection.Metadata.  When installing this into a PCL project I'll see 19 warnings which are not applicable since none of the downgrades actually have an effect.

```
{
  "supports": {
    "net46.app": {},
    "uwp.10.0.app": {},
    "dnxcore50.app": {}
  },
  "dependencies": {
    "Microsoft.NETCore": "5.0.0",
    "Microsoft.NETCore.Portable.Compatibility": "1.0.0",
    "System.Reflection.Metadata": "1.0.22"
  },
  "frameworks": {
    "dotnet": {}
  }
}
```
The same is happening for System.Collections.Immutable, with 13 warnings.

/cc @yishaigalatzer

**Comment by yishaigalatzer** [comment]
We can take a look

**Comment by davidfowl** [comment]
I'm not sure there's a bug here. This looks as though things are working. You can fix the warning by putting the downgraded package in your project.json

**Comment by ericstj** [comment]
At that point all it does is make the warning go away, Nuget was already selecting the correct package despite the warning.  I don't think it's correct to warn the developer when there isn't actually a downgrade happening.

**Comment by yishaigalatzer** [comment]
Can you please fix the dot net packages? We can take a look at the warnings

**Comment by davidfowl** [comment]
I see, we'd need to check if that node was chosen after cousin resolution?

**Comment by ericstj** [comment]
I have fixed most of the packages however there are some that we should not fix.  These packages are completely PCL. There are platforms where their dependencies don't result in downgrade.  The problem is that DotNET dependency section contains both implementation and reference dependencies since there is no good way to split them.  We don't want to leak platform specific implementation of other contracts up to these pure PCL packages.  If we do it will actually restrict where they can run.

**Comment by yishaigalatzer** [comment]
Lou is suggesting to walk up the graph to the end instead of warning as soon as we see a downgrade.

I'm not familiar enough with the code to guide Pranav on this, David can you help?

**Comment by davidfowl** [comment]
Sent a PR for this. The change should be small. It needs more validation though but it seems to work well.

**Comment by emgarten** [comment]
Fixed with NuGet/NuGet3@8f8e2bd8343b1b27bd168d184f62390555333e1d
